cns/client/client.go

```diff
@@ -0,0 +1,402 @@
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/Azure/azure-container-networking/cns/restserver"
+	"github.com/Azure/azure-container-networking/cns/types"
+	"github.com/pkg/errors"
+)
+
+const (
+	contentTypeJSON = "application/json"
+	defaultBaseURL  = "http://localhost:10090"
+	// DefaultTimeout default timeout duration for CNS Client.
+	DefaultTimeout    = 5 * time.Second
+	headerContentType = "Content-Type"
+)
+
+var clientPaths = []string{
+	cns.GetNetworkContainerByOrchestratorContext,
+	cns.CreateHostNCApipaEndpointPath,
+	cns.DeleteHostNCApipaEndpointPath,
+	cns.RequestIPConfig,
+	cns.ReleaseIPConfig,
+	cns.PathDebugIPAddresses,
+	cns.PathDebugPodContext,
+	cns.PathDebugRestData,
+}
+
+// Client specifies a client to connect to Ipam Plugin.
+type Client struct {
+	client http.Client
+	routes map[string]url.URL
+}
+
+// New returns a new CNS client configured with the passed URL and timeout.
+func New(baseURL string, requestTimeout time.Duration) (*Client, error) {
+	if baseURL == "" {
+		baseURL = defaultBaseURL
+	}
+
+	routes, err := buildRoutes(baseURL, clientPaths)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		client: http.Client{
+			Timeout: requestTimeout,
+		},
+		routes: routes,
+	}, nil
+}
+
+func buildRoutes(baseURL string, paths []string) (map[string]url.URL, error) {
+	base, err := url.Parse(baseURL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse base URL %s", baseURL)
+	}
+
+	routes := map[string]url.URL{}
+	for _, path := range paths {
+		u := *base
+		pathURI, err := url.Parse(path)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to parse path URI %s", path)
+		}
+		u.Path = pathURI.Path
+		routes[path] = u
+	}
+
+	return routes, nil
+}
+
+// GetNetworkConfiguration Request to get network config.
+func (c *Client) GetNetworkConfiguration(ctx context.Context, orchestratorContext []byte) (*cns.GetNetworkContainerResponse, error) {
+	payload := cns.GetNetworkContainerRequest{
+		OrchestratorContext: orchestratorContext,
+	}
+
+	var body bytes.Buffer
+	if err := json.NewEncoder(&body).Encode(payload); err != nil {
+		return nil, &CNSClientError{
+			Code: types.UnexpectedError,
+			Err:  err,
+		}
+	}
+
+	u := c.routes[cns.GetNetworkContainerByOrchestratorContext]
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, u.String(), &body)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to build request")
+	}
+	req.Header.Set(headerContentType, contentTypeJSON)
+	res, err := c.client.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "http request failed")
+	}
+	defer res.Body.Close()
```

Comment:
One thing that should also be added to all of these network calls is making sure the entire response body is read before closing, otherwise tcp connections won't be reused efficiently:

```go
defer func() {
    _, _ = io.Copy(ioutil.Discard, res.Body)
    res.Body.Close()
}()
```